pkg/server/commands/listusers/list_users_rpc.go

```diff
@@ -0,0 +1,346 @@
+	pool.WithCancelOnError()
+	pool.WithMaxGoroutines(int(l.resolveNodeBreadthLimit))
+
+	for {
+		tupleKey, err := filteredIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+
+			return err
+		}
+
+		tupleKeyUser := tupleKey.GetUser()
+
+		userObject, userRelation := tuple.SplitObjectRelation(tupleKeyUser)
+
+		userObjectType, userObjectID := tuple.SplitObject(userObject)
+
+		if userRelation == "" {
+			for _, f := range req.GetUserFilters() {
+				if f.GetType() == userObjectType {
+					// we found one, time to return it!
+					foundUsersChan <- &openfgav1.User{
+						User: &openfgav1.User_Object{
+							Object: &openfgav1.Object{
+								Type: userObjectType,
+								Id:   userObjectID,
+							},
+						},
+					}
+				}
+			}
+			continue
+		}
+
+		pool.Go(func(ctx context.Context) error {
+
+			return l.expand(ctx, &openfgav1.ListUsersRequest{
+				StoreId:              req.GetStoreId(),
+				AuthorizationModelId: req.GetAuthorizationModelId(),
+				Object:               &openfgav1.Object{Type: userObjectType, Id: userObjectID},
+				Relation:             userRelation,
+				UserFilters:          req.GetUserFilters(),
+				ContextualTuples:     req.GetContextualTuples(),
+			}, foundUsersChan)
+		})
+
+	}
+
+	return pool.Wait()
+}
+
+func (l *listUsersQuery) expandTTU(
+	ctx context.Context,
+	req listUsersRequest,
+	rewrite *openfgav1.Userset_TupleToUserset,
+	foundUsersChan chan<- *openfgav1.User,
+) error {
+	tuplesetRelation := rewrite.TupleToUserset.GetTupleset().GetRelation()
+	computedRelation := rewrite.TupleToUserset.GetComputedUserset().GetRelation()
+
+	typesys, err := l.typesystemResolver(ctx, req.GetStoreId(), req.GetAuthorizationModelId())
+	if err != nil {
+		return err
+	}
+
+	iter, err := l.ds.Read(ctx, req.GetStoreId(), &openfgav1.TupleKey{
+		Object:   tuple.ObjectKey(req.GetObject()),
+		Relation: tuplesetRelation,
+	})
+	if err != nil {
+		return err
+	}
+	defer iter.Stop()
+
+	filteredIter := storage.NewFilteredTupleKeyIterator(
+		storage.NewTupleKeyIteratorFromTupleIterator(iter),
+		validation.FilterInvalidTuples(typesys),
+	)
+	defer filteredIter.Stop()
+
+	pool := pool.New().WithContext(ctx)
+	pool.WithCancelOnError()
+	pool.WithMaxGoroutines(int(l.resolveNodeBreadthLimit))
+
+	for {
+		tupleKey, err := filteredIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+
+			return err
+		}
+
+		userObject := tupleKey.GetUser()
+		userObjectType, userObjectID := tuple.SplitObject(userObject)
+
+		pool.Go(func(ctx context.Context) error {
+			return l.expand(ctx, &openfgav1.ListUsersRequest{
+				StoreId:              req.GetStoreId(),
+				AuthorizationModelId: req.GetAuthorizationModelId(),
+				Object:               &openfgav1.Object{Type: userObjectType, Id: userObjectID},
+				Relation:             computedRelation,
+				UserFilters:          req.GetUserFilters(),
+			}, foundUsersChan)
```

Comment:
We should talk over Zoom about this because it's hard to explain here, but: if you use `ListUsersRequest` to call `expand` recursively you will lose information about the graph traversal that you are doing. Specifically, you are losing info about which "node" you are currently at. 

So, similar to https://github.com/openfga/openfga/blob/7bb7fe40ab28610994cfdd75744edb0ae6523d8e/pkg/server/commands/reverseexpand/reverse_expand.go#L35-L44, you need to create a new type:

```go
var _ listUsersRequest = (*forwardExpandRequest)(nil)

type forwardExpandRequest struct {
	StoreID              string
	AuthorizationModelID string

	// node in the graph we are trying to reach (the ultimate target)
	UserFilters []*openfgav1.ListUsersFilter

	// node in the graph we will visit in this expansion (i.e. FROM edge = head of edge)
	Object   *openfgav1.Object
	Relation string

	// the edge we are visiting in this expansion (includes info about the TO of the edge = tail of edge)
	Edge *graph.RelationshipEdge
}
```

The only moment when you can say for sure that you found an user is when the tail of the edge you are currently visiting == the filter, i.e.

```
if req.edge.TargetReference.GetType() == f.GetType() && req.edge.TargetReference.GetRelation() == f.GetRelation() {
	foundUsersChan <- &openfgav1.User{
		User: &openfgav1.User_Object{
			Object: req.Object,
			},
		}
}
```